src/validate.py

```diff
@@ -0,0 +1,157 @@
+
+def main() -> None:
+    """CLI."""
+    parser = argparse.ArgumentParser(
+                prog="CSV file validation",
+                description="Validates the formatting and typing of csv files."
+             )
+    
+    parser.add_argument(
+            "-d",
+            "--data",
+            help="The path to the data folder containing the csv files.",
+            type=Path,
+            required=True
+            )
+
+    parser.add_argument(
+            "-c",
+            "--configfile",
+            help="The path to the validation.toml.",
+            type=Path,
+            default=PATH.parent.parent / "data" / "metadata_test" / "_validation_schema_v2.toml"
+            )
+
+    parser.add_argument(
+            "-s",
+            "--sep",
+            help="Separator between columns in csv files. \n Tab separator: tab. Default: ;",
+            type=str,
+            default=";"
+            )
+
+    parser.add_argument(
+            "-v",
+            "--verbose",
+            help="Get more output in the single steps of the pipeline to debug.",
+            action='store_true'
+            )
+
+    args, _ = parser.parse_known_args() 
+   
+    # For tab separated files
+    if args.sep == "tab":
+        args.sep = "\t"
+
+    # read in toml file
+    validation = read_toml(args.configfile)
+    if args.verbose == True:
+        print(f"DEBUG: validation.toml.")
+        pprint.pprint(validation)
+    # read in data
+    data = read_csv_files(args.data, args.sep)
+    if args.verbose == True:
+        for df_name in data:
+            print(f"DEBUG:shape of {df_name}: {data[df_name].shape}")
+    print("Read in csv files done.")
+
+    # check if all columns exist
+    result = check_column_exists(data, validation)
+    counter = 0
+    for fname, col in result:
+        print(f"{FAIL}Missing column {col} in {fname}.{ENDC}")
+        counter+=1
+        if counter == MAX_NUM_RES-1:
+            print(f"First {MAX_NUM_RES} results out of {len(result)}")
+            break
+    print("Check if all columns are present in the csv completed.")
+
+    # check column types
+    result = check_column_types(data, validation)
+    counter = 0
+    for fname, col, fount_type, exp_type in result:
+        print(f"{FAIL}TYPE in {fname} column {col}; expected {exp_type}, found {fount_type}.{ENDC}")
+        counter+=1
+        if counter == MAX_NUM_RES-1:
+            print(f"First {MAX_NUM_RES} results out of {len(result)}")
+            break
+    print("Check column types completed.")
+
+    # checks on identifier columns
+    ids = ["host_id", "environment_id"]
+    if identifier_checks(data, ids):
+        print("Identifier columns checked successfully.")
+    else:
+        print("Errors found in identifer columns. Please correct them.")
+        return(-1)
+
+    # check dependencies between columns
+    for data_name in data:
+        if set(validation["column_dependencies"]["event"]).issubset(data[data_name].columns):
```

Comment:
I'm trying to follow the logic, but I'm really confused. From what I understand we say: If "event_day", "event_time", "event_type" are columns in the dataframe, then we need one of the three lists of columns (called "measurement", "inoculation", "treatment") to be present. From the schema, I have the feeling this is already enforced by the schema itself in [events] and [environment_events], or am I missing something? Also this logic is not clear from the schema, since "event" is on the same level as "measurement". So at least some more documentation in both the toml and validation itself is neccessary. I also think that this logic should go in the validation_utils, since the same logic is copied in the notebook.